addons/website/models/ir_qweb.py

```diff
@@ -20,6 +20,12 @@ def _get_asset_url_values(self, id, unique, extra, name, page, type):
         res = super(AssetsBundleMultiWebsite, self)._get_asset_url_values(id, unique, extra, name, page, type)
         return res
 
+    def _get_assets_domain_for_already_processed_css(self, assets):
+        res = super(AssetsBundleMultiWebsite, self)._get_assets_domain_for_already_processed_css(assets)
+        current_website = self.env['website'].get_current_website()
+        if current_website:
```

Comment:
@jpp-odoo Also, @rdeodoo is pretty annoying but he says you could use `expression.AND` to make the domain combination even more beautiful :smile: :+1:

Reply:
@qsm-odoo  ![giphy](https://user-images.githubusercontent.com/30048408/67020131-005d8880-f0fe-11e9-9afa-3980e54a0609.gif)

Reply:
The goal is not to be sure if we are in a frontend context but to find the assets with the correct website. As in the creation of the attachment https://github.com/odoo/odoo/blob/12.0/addons/website/models/ir_attachment.py#L22 it uses get_current_website with fallback=False, I think I'm going to do the same here, to be coherent with the creation of the element we search. 

What do you think ?

Also, I change to expression.AND, indeed its more beautiful :)

Reply:
Actually the line you mention will already lead to a misbehavior (which create no bug fortunately, for now).
It assumes `get_current_website(fallback=False)` will not return a website if we are not in a frontend context, but that's not always true.

If you check the code, the function will **always** go through the last code of the method if there is 2 websites with the same domain (which might be a real case if the user is using GEOIP multiwebsite):
```python
else:  # > 1 website with the same domain
    country_specific_websites = websites.filtered(lambda website: country_id in 
    website.country_group_ids.mapped('country_ids').ids)
    return country_specific_websites[0].id if country_specific_websites else websites[0].id
```
Then from that point, it will return you the first website, despite `fallback=True`. Every attachment created will now have a website_id, regardless if you are in a frontend or backend. (Not true for records created manually through the form view as they will have website_id set to False in the form view so the code line you provided will not go in the if condition).
I am pretty sure if the assets are rebuilt in that case, they will have the first website_id set on it, even if rebuilt in a backend context.

I am just feeling it coming that `current_website = self.env['website'].get_current_website(fallback=False)` will not return what you expect in every case.
Where `website = ir_http.get_request_website()` will. Either `False` will be returned if you are in the backend, or the website_id if you are in the frontend.

I do believe `get_current_website()` is never supposed to be used on code which might be called by the backend (@JKE-be am I wrong?).

I did not check further but whatever option you choose, you might want to check it works as you expect in both backend and frontend context.

Reply:
Hi @rdeodoo 

> I am pretty sure if the assets are rebuilt in that case, they will have the first website_id set on it, even if rebuilt in a backend context.

ok if it happen, but then this was an issue in the `create` before https://github.com/odoo/odoo/commit/42eec88e5da164e4 that is being reverted in this commit.

so as far as I understand there is 3 solutions:

### 1) current code

should work because when reading the attachment we also use `get_current_website`:

https://github.com/odoo/odoo/blob/f157234aa31b88ee6a90f37520116f31e0586d14/addons/website/models/ir_attachment.py#L32-L36

the only issue that I see is that backend assets will possibly be duplicated.

### 2) current code and get_current_website that works as it seem to be expected

ok

### 3) what you ask for:

```diff
--- a/addons/website/models/ir_qweb.py
+++ b/addons/website/models/ir_qweb.py
@@ -5,6 +5,7 @@
 
 from odoo import models
 from odoo.addons.http_routing.models.ir_http import url_for
+from odoo.addons.website.models import ir_http
+from odoo.osv import expression
 from odoo.http import request
 from odoo.addons.base.models.assetsbundle import AssetsBundle
 from odoo.tools import html_escape as escape
@@ -20,6 +21,12 @@ def _get_asset_url_values(self, id, unique, extra, name, page, type):
         res = super(AssetsBundleMultiWebsite, self)._get_asset_url_values(id, unique, extra, name, page, type)
         return res
 
+    def _get_assets_domain_for_already_processed_css(self, assets):
+        res = super(AssetsBundleMultiWebsite, self)._get_assets_domain_for_already_processed_css(assets)
+        current_website = ir_http.get_request_website()
+        return expression.AND([res, [('website_id', '=', False)]])
--- a/addons/website/models/ir_attachment.py
+++ b/addons/website/models/ir_attachment.py
@@ -3,6 +3,7 @@
 
 import logging
 from odoo import fields, models, api
+from odoo.addons.website.models import ir_http
 from odoo.exceptions import UserError
 from odoo.tools.translate import _
 _logger = logging.getLogger(__name__)
@@ -19,7 +20,7 @@ class Attachment(models.Model):
 
     @api.model
     def create(self, vals):
-        website = self.env['website'].get_current_website(fallback=False)
+        website = ir_http.get_request_website()
         if website and 'website_id' not in vals and 'not_force_website_id' not in self.env.context:
             vals['website_id'] = website.id
         return super(Attachment, self).create(vals)
@@ -30,14 +31,14 @@ class Attachment(models.Model):
 
     @api.model
     def get_serve_attachment(self, url, extra_domain=None, extra_fields=None, order=None):
-        website = self.env['website'].get_current_website()
+        website = ir_http.get_request_website()
         extra_domain = (extra_domain or []) + website.website_domain()
         order = ('website_id, %s' % order) if order else 'website_id'
         return super(Attachment, self).get_serve_attachment(url, extra_domain, extra_fields, order)
 
     @api.model
     def get_attachment_by_key(self, key, extra_domain=None, order=None):
-        website = self.env['website'].get_current_website()
+        website = ir_http.get_request_website()
         extra_domain = (extra_domain or []) + website.website_domain()
         order = ('website_id, %s' % order) if order else 'website_id'
         return super(Attachment, self).get_attachment_by_key(key, extra_domain, order)
```

But maybe there was a reason for `get_current_website` so this seems the most risky.

So I would be ok with solution 1 and 2, for solution 3 it seems more risky.

Reply:
I don't remember everything that has been discussed here and in the office.

But this code
```python
current_website = self.env['website'].get_current_website(fallback=False)
if current_website:
```
Is ringing a bell, it is basically a `if True` in some multiwebsite configs.
You never write a `if True`, do you?
Note that this is just something to triple check, I gave hints/warnings, I am not 100% sure it will and I might be wrong but I am pretty sure I am not.

But clearly, `get_current_website()` (even with fallback=False) will always return a website, even if called in backend, in some multiwebsite configuration. So gettings assets in backend will fetch website ones.

So, I am not confortable with `get_current_website()` called in method shared with backend. When called from frontend, it will most likely have `website_id` in context and break early from `get_current_website()` with the correct website.
It should ideally only be called in controller and test with `MockRequest` or something like that to simulate frontend context on backend/shared method.

If I get this one, we want to identify the website we are getting the assets for, no?
So: backend (no website) or frontend (find which website).
This is typically done with `get_request_website()` with 100% accuracy.

Reply:
then it is solution 3, as a side note here are the difference as far as I can tell:

#### before https://github.com/odoo/odoo/commit/42eec88e5da164e4 (october 2019):

=> in backend: we will create a new assets with `website_id=get_current_website()` if there is no assets with the given URL
=> in frontend: we will create a new assets with `website_id=get_current_website()` if there is no assets with the given URL

==> this does not work at all in multi website: asset does not exist on other website

#### after https://github.com/odoo/odoo/commit/42eec88e5da164e4:

=> in backend: we will create a new assets with `website_id=False` if there is no assets with the given URL
=> in frontend: we will create a new assets with `website_id=False` if there is no assets with the given URL

==> this does not work in multi website: the asset of only one website will be kept even if they differ (which seems to be possible)

#### after this current pr:

=> in backend: we will create a new assets with `website_id=get_current_website()` if there is no assets with the given URL that match `website_id=False` or `website_id=get_current_website()`
=> in frontend: we will create a new assets with `website_id=get_current_website()` if there is no assets with the given URL that match `website_id=False` or `website_id=get_current_website()`

==> this could work: in your "IF TRUE" case the backend asset will have a `website_id` but it will also be used to read it (because we read assets of get_current_website() so it is two times wrong which make a right)

#### after this current pr (solution 2 in my message):

=> OK but might be more side-effects we did not consider

#### after this current pr (solution 3 in my message):

=> OK but might be mpre side-effects we did not consider